### PR TITLE
:bug: Remove dataImage finalizer if BMH is missing

### PR DESCRIPTION
**Pull request by hroyrh**
This fixes the issue of dataImage resource not being deleted, if the corresponding BMH is deleted without detaching the dataImage.

### CODE REVIEW

**Review by eranco74 — COMMENTED** [review]
/lgtm

**Comment by dtantsur** [comment]
This is an interesting edge case. Is it because of a forced deletion of the BMH?

**Comment by hroyrh** [comment]
> This is an interesting edge case. Is it because of a forced deletion of the BMH?

Do you mean removing BMH's finalizer ? I tested with both - removing the finalizer and waiting for BMH to deprovision, after triggering BMH delete by running `kubectl delete bmh node-name`. In both cases, the DataImage was left orphaned.

**Comment by dtantsur** [comment]
> waiting for BMH to deprovision, after triggering BMH delete by running kubectl delete bmh node-name

At least this case should be handled by the BMH controller, not the DataImage controller, I think. ALso, don't we disconnect the ISO on a "normal" deprovisioning?

**Comment by hroyrh** [comment]
> At least this case should be handled by the BMH controller, not the DataImage controller, I think. ALso, don't we disconnect the ISO on a "normal" deprovisioning?

Why does that make more sense, won't a force deletion of BMH ( by removing the finalizer ) leave the dataImage orphaned in that case ?
I need to verify whether ISO disconnection happens on a "normal" deprovisioning.

**Comment by hroyrh** [comment]
> Your PR handles forced deletion, but I don't think it handles the normal deprovisioning case? The BMH controller must decide what to do with DataImages it owns. Either remove the finalizer (esp. when the node is detached) or disconnect the ISO first?

Makes sense, I will work on the detached and normal delete process. What about the forced deletion scenario, should I keep the current DataImage controller fix in place ?

**Comment by eranco74** [comment]
> ested with both - removing the finalizer and waiting for BMH to deprovision, after triggering BMH delete by running `kub`

not forced, BMH deletion isn't blocked on dataImage existence or attachment.
You can delete a BMH attached to dataImage, but then you end up with a dataImage you can't delete

**Comment by dtantsur** [comment]
> What about the forced deletion scenario, should I keep the current DataImage controller fix in place ?

I think it's a fair to remove the finalizer if the controller sees that the owner BMH is gone.

**Comment by mquhuy** [comment]
lgtm, but could we have a couple of e2e tests related to this?

**Comment by iurygregory** [comment]
lgtm, maybe a separate PR with the e2e would be ok (I do think it would be important)

**Comment by hroyrh** [comment]
Right, I can add the e2e tests in a separate PR.

**Comment by elfosardo** [comment]
/approve
let's add the e2e tests as soon as this merges

**Comment by tuminoid** [comment]
> Right, I can add the e2e tests in a separate PR.

https://github.com/metal3-io/baremetal-operator/issues/2053 so they're not forgotten, or at least are tracked.

File: controllers/metal3.io/baremetalhost_controller.go
Comment (dtantsur) [inline controllers/metal3.io/baremetalhost_controller.go:531]: Why do we ignore the error here? I'm fine with ignoring HTTP 404 (in case somebody did a nasty force-deletion), but it seems that ignoring any errors may result with orphaned DataImage.

File: controllers/metal3.io/baremetalhost_controller.go
Comment (dtantsur) [inline controllers/metal3.io/baremetalhost_controller.go:609]: Same question.

Also, %w does nothing, there is no string interpolation here.

File: controllers/metal3.io/baremetalhost_controller.go
Comment (dtantsur) [inline controllers/metal3.io/baremetalhost_controller.go:1437]: Won't this cause the image attachment to also happen in the available state?

File: controllers/metal3.io/baremetalhost_controller.go
Comment (dtantsur) [inline controllers/metal3.io/baremetalhost_controller.go:1516]: nit: unnecessary variable

File: controllers/metal3.io/baremetalhost_controller.go
Comment (dtantsur) [inline controllers/metal3.io/baremetalhost_controller.go:1438]: nit: I'm not sure we can enter this function in any other state.

File: controllers/metal3.io/baremetalhost_controller.go
Comment (dtantsur) [inline controllers/metal3.io/baremetalhost_controller.go:1517]: nit: I'd prefer this function to have an argument like `detachOnly bool` and keep the provision state logic at the caller side (where you already check for the Available state).

File: controllers/metal3.io/baremetalhost_controller.go
Comment (dtantsur) [inline controllers/metal3.io/baremetalhost_controller.go:1598]: The %w part will be logged verbatim, there is no string interpolation here.

File: controllers/metal3.io/baremetalhost_controller.go
Comment (dtantsur) [reply controllers/metal3.io/baremetalhost_controller.go:1438]: Thinking more about it, I don't believe using Available is the right thing to do. The image won't be disconnected until a reboot, which will happen.. who knows when. I think we should start deprovisioning with deleting a data image.

File: controllers/metal3.io/baremetalhost_controller.go
Comment (dtantsur) [inline controllers/metal3.io/baremetalhost_controller.go:1591]: Are we going to leave a stale DataImage object in this case? If yes, it will behave unpredictably on the next deployment because it will still have a non-empty URL.

**Review by dtantsur — CHANGES_REQUESTED** [review]
Some comments inline.

On top of that, I think this PR tries to solve at least 3 related but distinct problem (removing DataImage on detachment, on deletion, fixing orphaned DataImage objects). I'd rather have 3 PRs.

File: controllers/metal3.io/baremetalhost_controller.go
Comment (hroyrh) [reply controllers/metal3.io/baremetalhost_controller.go:1438]: You mean just deleting the data image and not detaching it, right ?

File: controllers/metal3.io/baremetalhost_controller.go
Comment (hroyrh) [reply controllers/metal3.io/baremetalhost_controller.go:1517]: Makes sense, I will make the changes.

File: controllers/metal3.io/baremetalhost_controller.go
Comment (hroyrh) [reply controllers/metal3.io/baremetalhost_controller.go:1591]: Good point, I will also set the `spec.url: ""`.

**Review by dtantsur — APPROVED** [review]

**Pull request merged**
